nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-standard-record-utils/src/main/java/org/apache/nifi/schema/validation/StandardSchemaValidator.java

```diff
@@ -233,41 +247,57 @@ private boolean isTypeCorrect(final Object value, final DataType dataType) {
                 }
 
                 return false;
-            case BIGINT:
-                return value instanceof BigInteger;
             case BOOLEAN:
                 return value instanceof Boolean;
-            case BYTE:
-                return value instanceof Byte;
             case CHAR:
                 return value instanceof Character;
             case DATE:
                 return value instanceof java.sql.Date;
-            case DOUBLE:
-                return value instanceof Double;
-            case FLOAT:
-                // Some readers do not provide float vs. double.
-                // We should consider if it makes sense to allow either a Float or a Double here or have
-                // a Reader indicate whether or not it supports higher precision, etc.
-                // Same goes for Short/Integer
-                return value instanceof Float;
-            case INT:
-                return value instanceof Integer;
-            case LONG:
-                return value instanceof Long;
-            case SHORT:
-                return value instanceof Short;
             case STRING:
                 return value instanceof String;
             case TIME:
                 return value instanceof java.sql.Time;
             case TIMESTAMP:
                 return value instanceof java.sql.Timestamp;
+
+            // Numeric data types
+            case BIGINT:
+            case LONG:
+            case INT:
+            case SHORT:
+            case BYTE:
+            case DOUBLE:
+                return isFittingNumberType(value, dataType.getFieldType());
+            case FLOAT:
+                // Some readers do not provide float vs. double.
+                // We should consider if it makes sense to allow either a Float or a Double here or have
+                // a Reader indicate whether or not it supports higher precision, etc.
+                // Same goes for Short/Integer
+                return isFittingNumberType(value, dataType.getFieldType()) || isDoubleWithinFloatInterval(value);
         }
 
         return false;
     }
 
+    /**
+     * Checks if an incoming value satisfies the requirements of a given (numeric) type or any of it's narrow data type.
+     *
+     * @param value Incoming value.
+     * @param fieldType The expected field type.
+     *
+     * @return Returns true if the incoming value satisfies the data type of any of it's narrow data types. Otherwise returns false. Only numeric data types are supported.
+     */
+    private boolean isFittingNumberType(final Object value, final RecordFieldType fieldType) {
```

Comment:
This is a really expensive method. This is going to be run against potentially every record field in a Record, so in some cases millions of times per second. We should ensure that it's extremely efficient:
- Use of `containsKey` and `get` shouldn't be used. Since we know that the Map will never contain `null` values, a simple `get` should be used, and then null-checked.
- getNarrowDataTypes() is fine, but calling `stream()` is very expensive. Should use a for-each instead.
- For each value in the stream, it's doing another HashMap lookup. This is relatively expensive.

To be honest, I would lean toward avoiding this hashmap lookup at all, and instead just update the case statements to be something along the lines of:
```
case LONG:
    return value instanceof Long || value instanceof Integer || value instanceof Short || value instanceof Byte;
```
Not saying that you've gotta go this way, but it would result in code that is very simple and easy to understand, and very efficient as well.

Reply:
I would try to keep away from double booking "narrow types" and still building on top of the list maintained in the given types on RecordFieldType. Assuming that instead on `HashMap` we would use `EnumMap` (I do not even know why I did not use this in the first place) what do you think about the following version:

```
 private boolean isFittingNumberType(final Object value, final RecordFieldType fieldType) {
        if (NUMERIC_VALIDATORS.get(fieldType).test(value)) {
            return true;
        }

        for (final RecordFieldType recordFieldType : fieldType.getNarrowDataTypes()) {
            if (NUMERIC_VALIDATORS.get(recordFieldType).test(value)) {
                return true;
            }
        }

        return false;
```

I think this aims the issues you point out. If your opinion is still that it's not effective, of course I can go on with your suggestion.